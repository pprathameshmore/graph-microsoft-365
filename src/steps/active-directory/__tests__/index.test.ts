import {
  createMockStepExecutionContext,
  Recording,
} from '@jupiterone/integration-sdk-testing';
import { entities, fetchAccount } from '..';
import {
  config,
  insufficientPermissionsDirectoryConfig,
  noMdmConfig,
} from '../../../../test/config';
import { setupAzureRecording } from '../../../../test/recording';

let recording: Recording;

afterEach(async () => {
  if (recording) {
    await recording.stop();
  }
});

describe('fetchAccount', () => {
  it('Should create an account entity correctly when the account has the correct permissions', async () => {
    recording = setupAzureRecording({
      directory: __dirname,
      name: 'fetchAccount',
    });
    const context = createMockStepExecutionContext({ instanceConfig: config });

    await fetchAccount(context);

    const accountEntities = context.jobState.collectedEntities;

    expect(accountEntities.length).toBe(1);
    expect(accountEntities).toMatchGraphObjectSchema({
      _class: entities.ACCOUNT._class,
    });
    expect(accountEntities).toMatchSnapshot('accountEntitiesSuccessful');
  });
  it('Should error when there are errors attempting to get organization data', () => {
    recording = setupAzureRecording({
      directory: __dirname,
      name: 'fetchAccountFail',
      options: {
        recordFailedRequests: true, // getting organization data will fail
      },
    });
    const context = createMockStepExecutionContext({
      instanceConfig: insufficientPermissionsDirectoryConfig,
    });
    expect(fetchAccount(context)).rejects;
  });

  it('Should not error and create a real account when there is no mdm authority', async () => {
    recording = setupAzureRecording({
      directory: __dirname,
      name: 'fetchAccountNoMdm',
      options: {
        recordFailedRequests: true, // getting the intune subscription will fail
      },
    });
    const context = createMockStepExecutionContext({
      instanceConfig: noMdmConfig,
    });

    await fetchAccount(context);

    const accountEntities = context.jobState.collectedEntities;

    expect(accountEntities.length).toBe(1);
    expect(accountEntities).toMatchGraphObjectSchema({
      _class: entities.ACCOUNT._class,
    });
    expect(accountEntities).toMatchSnapshot('accountEntitiesNoMdm');
  });
});
