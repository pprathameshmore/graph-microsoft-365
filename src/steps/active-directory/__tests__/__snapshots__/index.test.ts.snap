// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchAccount Should create an account entity correctly when the account has the correct permissions: accountEntitiesSuccessful 1`] = `
Array [
  Object {
    "_class": Array [
      "Account",
    ],
    "_key": "microsoft_365_account-local-integration-instance",
    "_rawData": Array [
      Object {
        "name": "default",
        "rawData": Object {
          "intuneConfig": Object {
            "intuneAccountID": "7b521133-f056-40a9-b639-b76afd67e4bb",
            "mobileDeviceManagementAuthority": "intune",
            "subscriptionState": "active",
          },
          "organization": Object {
            "@odata.id": "https://graph.microsoft.com/v2/a76fc728-0cba-45f0-a9eb-d45207e14513/directoryObjects/a76fc728-0cba-45f0-a9eb-d45207e14513/Microsoft.DirectoryServices.Organization",
            "assignedPlans": Array [
              Object {
                "assignedDateTime": "2021-03-15T14:57:06Z",
                "capabilityStatus": "Suspended",
                "service": "MicrosoftKaizala",
                "servicePlanId": "54fc630f-5a40-48ee-8965-af0503c1386e",
              },
              Object {
                "assignedDateTime": "2021-03-15T14:57:06Z",
                "capabilityStatus": "Suspended",
                "service": "CRM",
                "servicePlanId": "4ff01e01-1ba7-4d71-8cf8-ce96c3bbcf14",
              },
              Object {
                "assignedDateTime": "2021-03-15T14:57:06Z",
                "capabilityStatus": "Suspended",
                "service": "exchange",
                "servicePlanId": "33c4f319-9bdd-48d6-9c4d-410b750a4a5a",
              },
              Object {
                "assignedDateTime": "2021-03-15T14:57:07Z",
                "capabilityStatus": "Suspended",
                "service": "MicrosoftCommunicationsOnline",
                "servicePlanId": "0feaeb32-d00e-4d66-bd5a-43b5b83db82c",
              },
              Object {
                "assignedDateTime": "2021-03-15T14:57:06Z",
                "capabilityStatus": "Suspended",
                "service": "CRM",
                "servicePlanId": "95b76021-6a53-4741-ab8b-1d1f3d66a95a",
              },
              Object {
                "assignedDateTime": "2021-03-15T14:57:06Z",
                "capabilityStatus": "Suspended",
                "service": "ProjectProgramsAndPortfolios",
                "servicePlanId": "31b4e2fc-4cd6-4e7d-9c1b-41407303bd66",
              },
              Object {
                "assignedDateTime": "2021-03-15T14:57:06Z",
                "capabilityStatus": "Suspended",
                "service": "MicrosoftStream",
                "servicePlanId": "3c53ea51-d578-46fa-a4c0-fd0a92809a60",
              },
              Object {
                "assignedDateTime": "2021-03-15T14:57:07Z",
                "capabilityStatus": "Suspended",
                "service": "SharePoint",
                "servicePlanId": "e95bec33-7c88-4a70-8e19-b10bd9d0c014",
              },
              Object {
                "assignedDateTime": "2021-03-15T14:57:06Z",
                "capabilityStatus": "Suspended",
                "service": "ProcessSimple",
                "servicePlanId": "0f9b09cb-62d1-4ff4-9129-43f4996f83f4",
              },
              Object {
                "assignedDateTime": "2021-03-15T14:57:06Z",
                "capabilityStatus": "Suspended",
                "service": "OfficeForms",
                "servicePlanId": "159f4cd6-e380-449f-a816-af1a9ef76344",
              },
              Object {
                "assignedDateTime": "2021-03-15T14:57:07Z",
                "capabilityStatus": "Suspended",
                "service": "YammerEnterprise",
                "servicePlanId": "7547a3fe-08ee-4ccb-b430-5077c5041653",
              },
              Object {
                "assignedDateTime": "2021-03-15T14:57:07Z",
                "capabilityStatus": "Suspended",
                "service": "exchange",
                "servicePlanId": "5bfe124c-bbdc-4494-8835-f1297d457d79",
              },
              Object {
                "assignedDateTime": "2021-03-15T14:57:06Z",
                "capabilityStatus": "Suspended",
                "service": "ccibotsprod",
                "servicePlanId": "041fe683-03e4-45b6-b1af-c0cdc516daee",
              },
              Object {
                "assignedDateTime": "2021-03-15T14:57:06Z",
                "capabilityStatus": "Suspended",
                "service": "exchange",
                "servicePlanId": "199a5c09-e0ca-4e37-8f7c-b05d533e1ea2",
              },
              Object {
                "assignedDateTime": "2021-03-15T14:57:07Z",
                "capabilityStatus": "Suspended",
                "service": "exchange",
                "servicePlanId": "9aaf7827-d63c-4b61-89c3-182f06f82e5c",
              },
              Object {
                "assignedDateTime": "2021-03-15T14:57:06Z",
                "capabilityStatus": "Suspended",
                "service": "To-Do",
                "servicePlanId": "5e62787c-c316-451f-b873-1d05acd4d12c",
              },
              Object {
                "assignedDateTime": "2021-03-15T14:57:06Z",
                "capabilityStatus": "Suspended",
                "service": "WhiteboardServices",
                "servicePlanId": "b8afc642-032e-4de5-8c0a-507a7bba7e5d",
              },
              Object {
                "assignedDateTime": "2021-03-15T14:57:06Z",
                "capabilityStatus": "Suspended",
                "service": "Deskless",
                "servicePlanId": "8c7d2df8-86f0-4902-b2ed-a0458298f3b3",
              },
              Object {
                "assignedDateTime": "2021-03-15T14:57:07Z",
                "capabilityStatus": "Suspended",
                "service": "MicrosoftOffice",
                "servicePlanId": "094e7854-93fc-4d55-b2c0-3ab5369ebdc1",
              },
              Object {
                "assignedDateTime": "2021-03-15T14:57:06Z",
                "capabilityStatus": "Suspended",
                "service": "DynamicsNAV",
                "servicePlanId": "39b5c996-467e-4e60-bd62-46066f572726",
              },
              Object {
                "assignedDateTime": "2021-03-15T14:57:07Z",
                "capabilityStatus": "Suspended",
                "service": "ProjectWorkManagement",
                "servicePlanId": "b737dad2-2f6c-4c65-90e3-ca563267e8b9",
              },
              Object {
                "assignedDateTime": "2021-03-15T14:57:06Z",
                "capabilityStatus": "Suspended",
                "service": "Bing",
                "servicePlanId": "94065c59-bc8e-4e8b-89e5-5138d471eaff",
              },
              Object {
                "assignedDateTime": "2021-03-15T14:57:07Z",
                "capabilityStatus": "Suspended",
                "service": "SCO",
                "servicePlanId": "882e1d05-acd1-4ccb-8708-6ee03664b117",
              },
              Object {
                "assignedDateTime": "2021-03-15T14:57:07Z",
                "capabilityStatus": "Suspended",
                "service": "PowerAppsService",
                "servicePlanId": "92f7a6f3-b89b-4bbd-8c30-809e6da5ad1c",
              },
              Object {
                "assignedDateTime": "2021-03-15T14:57:07Z",
                "capabilityStatus": "Suspended",
                "service": "Sway",
                "servicePlanId": "a23b959c-7ce8-4e57-9140-b90eb88a9e97",
              },
              Object {
                "assignedDateTime": "2021-03-15T14:57:06Z",
                "capabilityStatus": "Suspended",
                "service": "RMSOnline",
                "servicePlanId": "31cf2cfc-6b0d-4adc-a336-88b724ed8122",
              },
              Object {
                "assignedDateTime": "2021-03-15T14:57:07Z",
                "capabilityStatus": "Suspended",
                "service": "TeamspaceAPI",
                "servicePlanId": "57ff2da0-773e-42df-b2af-ffb7a2317929",
              },
              Object {
                "assignedDateTime": "2021-03-15T14:57:07Z",
                "capabilityStatus": "Suspended",
                "service": "SharePoint",
                "servicePlanId": "c7699d2e-19aa-44de-8edf-1736da088ca1",
              },
              Object {
                "assignedDateTime": "2021-02-15T15:32:43Z",
                "capabilityStatus": "Enabled",
                "service": "exchange",
                "servicePlanId": "113feb6c-3fe4-4440-bddc-54d774bf0318",
              },
              Object {
                "assignedDateTime": "2021-02-15T15:32:43Z",
                "capabilityStatus": "Enabled",
                "service": "Adallom",
                "servicePlanId": "2e2ddb96-6af9-4b1d-a3f0-d6ecfd22edb2",
              },
              Object {
                "assignedDateTime": "2021-02-15T15:32:43Z",
                "capabilityStatus": "Enabled",
                "service": "AADPremiumService",
                "servicePlanId": "eec0eb4f-6444-4f95-aba0-50c24d67f998",
              },
              Object {
                "assignedDateTime": "2021-02-15T15:32:43Z",
                "capabilityStatus": "Enabled",
                "service": "AADPremiumService",
                "servicePlanId": "41781fb2-bc02-4b7c-bd55-b576c07bb09d",
              },
              Object {
                "assignedDateTime": "2021-02-15T15:32:43Z",
                "capabilityStatus": "Enabled",
                "service": "RMSOnline",
                "servicePlanId": "6c57d4b6-3b23-47a5-9bc9-69f17b4947b3",
              },
              Object {
                "assignedDateTime": "2021-02-15T15:32:43Z",
                "capabilityStatus": "Enabled",
                "service": "RMSOnline",
                "servicePlanId": "bea4c11e-220a-4e6d-8eb8-8ea15d019f90",
              },
              Object {
                "assignedDateTime": "2021-02-15T15:32:43Z",
                "capabilityStatus": "Enabled",
                "service": "SCO",
                "servicePlanId": "c1ec4a95-1f05-45b3-a911-aa3fa01094f5",
              },
              Object {
                "assignedDateTime": "2021-02-15T15:32:43Z",
                "capabilityStatus": "Enabled",
                "service": "RMSOnline",
                "servicePlanId": "5689bec4-755d-4753-8b61-40975025187c",
              },
              Object {
                "assignedDateTime": "2021-02-15T15:32:43Z",
                "capabilityStatus": "Enabled",
                "service": "MultiFactorService",
                "servicePlanId": "8a256a2b-b617-496d-b51b-e76466e88db0",
              },
              Object {
                "assignedDateTime": "2021-02-15T15:32:43Z",
                "capabilityStatus": "Enabled",
                "service": "AzureAdvancedThreatAnalytics",
                "servicePlanId": "14ab5db5-e6c4-4b20-b4bc-13e36fd2227f",
              },
              Object {
                "assignedDateTime": "2021-02-13T14:57:07Z",
                "capabilityStatus": "Deleted",
                "service": "SCO",
                "servicePlanId": "882e1d05-acd1-4ccb-8708-6ee03664b117",
              },
              Object {
                "assignedDateTime": "2019-09-04T17:04:05Z",
                "capabilityStatus": "Enabled",
                "service": "WindowsAzure",
                "servicePlanId": "fca3e605-0754-4279-8504-3f1229f29614",
              },
            ],
            "businessPhones": Array [],
            "city": null,
            "country": null,
            "countryLetterCode": "US",
            "createdDateTime": "2019-09-04T16:59:36Z",
            "deletedDateTime": null,
            "directorySizeQuota": Object {
              "total": 50000,
              "used": 569,
            },
            "displayName": "Default Directory",
            "id": "a76fc728-0cba-45f0-a9eb-d45207e14513",
            "isMultipleDataLocationsForServicesEnabled": null,
            "marketingNotificationEmails": Array [],
            "onPremisesLastSyncDateTime": null,
            "onPremisesSyncEnabled": null,
            "postalCode": null,
            "preferredLanguage": "en",
            "privacyProfile": null,
            "provisionedPlans": Array [
              Object {
                "capabilityStatus": "Suspended",
                "provisioningStatus": "Success",
                "service": "MicrosoftCommunicationsOnline",
              },
              Object {
                "capabilityStatus": "Suspended",
                "provisioningStatus": "Success",
                "service": "YammerEnterprise",
              },
              Object {
                "capabilityStatus": "Suspended",
                "provisioningStatus": "Success",
                "service": "SharePoint",
              },
              Object {
                "capabilityStatus": "Suspended",
                "provisioningStatus": "Success",
                "service": "SharePoint",
              },
              Object {
                "capabilityStatus": "Suspended",
                "provisioningStatus": "Success",
                "service": "RMSOnline",
              },
              Object {
                "capabilityStatus": "Suspended",
                "provisioningStatus": "Success",
                "service": "SCO",
              },
              Object {
                "capabilityStatus": "Suspended",
                "provisioningStatus": "Success",
                "service": "ProjectWorkManagement",
              },
              Object {
                "capabilityStatus": "Suspended",
                "provisioningStatus": "Success",
                "service": "exchange",
              },
              Object {
                "capabilityStatus": "Suspended",
                "provisioningStatus": "Success",
                "service": "exchange",
              },
              Object {
                "capabilityStatus": "Suspended",
                "provisioningStatus": "Success",
                "service": "exchange",
              },
              Object {
                "capabilityStatus": "Suspended",
                "provisioningStatus": "Success",
                "service": "exchange",
              },
              Object {
                "capabilityStatus": "Enabled",
                "provisioningStatus": "Success",
                "service": "Adallom",
              },
              Object {
                "capabilityStatus": "Enabled",
                "provisioningStatus": "Success",
                "service": "SCO",
              },
              Object {
                "capabilityStatus": "Enabled",
                "provisioningStatus": "Success",
                "service": "exchange",
              },
              Object {
                "capabilityStatus": "Enabled",
                "provisioningStatus": "Success",
                "service": "RMSOnline",
              },
              Object {
                "capabilityStatus": "Enabled",
                "provisioningStatus": "Success",
                "service": "RMSOnline",
              },
              Object {
                "capabilityStatus": "Enabled",
                "provisioningStatus": "Success",
                "service": "RMSOnline",
              },
            ],
            "securityComplianceNotificationMails": Array [],
            "securityComplianceNotificationPhones": Array [],
            "state": null,
            "street": null,
            "technicalNotificationMails": Array [
              "adam.jupiterone@hotmail.com",
            ],
            "tenantType": "AAD",
            "verifiedDomains": Array [
              Object {
                "capabilities": "Email, OfficeCommunicationsOnline",
                "isDefault": true,
                "isInitial": true,
                "name": "adamjupiteronehotmailcom.onmicrosoft.com",
                "type": "Managed",
              },
            ],
          },
        },
      },
    ],
    "_type": "microsoft_365_account",
    "createdOn": undefined,
    "defaultDomain": "adamjupiteronehotmailcom.onmicrosoft.com",
    "displayName": "Local Integration",
    "id": "a76fc728-0cba-45f0-a9eb-d45207e14513",
    "intuneAccountId": "7b521133-f056-40a9-b639-b76afd67e4bb",
    "intuneSubscriptionState": "active",
    "mobileDeviceManagementAuthority": "intune",
    "name": "Default Directory",
    "organizationName": "Default Directory",
    "verifiedDomains": Array [
      "adamjupiteronehotmailcom.onmicrosoft.com",
    ],
  },
]
`;

exports[`fetchAccount Should not error and create a real account when there is no mdm authority: accountEntitiesNoMdm 1`] = `
Array [
  Object {
    "_class": Array [
      "Account",
    ],
    "_key": "microsoft_365_account-local-integration-instance",
    "_rawData": Array [
      Object {
        "name": "default",
        "rawData": Object {
          "intuneConfig": Object {
            "intuneAccountID": undefined,
            "mobileDeviceManagementAuthority": undefined,
            "subscriptionState": undefined,
          },
          "organization": Object {
            "@odata.id": "https://graph.microsoft.com/v2/75f092de-c730-4a62-a1dd-4324cb0b83af/directoryObjects/75f092de-c730-4a62-a1dd-4324cb0b83af/Microsoft.DirectoryServices.Organization",
            "assignedPlans": Array [],
            "businessPhones": Array [],
            "city": null,
            "country": null,
            "countryLetterCode": "US",
            "createdDateTime": "2021-02-09T14:31:53Z",
            "deletedDateTime": null,
            "directorySizeQuota": Object {
              "total": 50000,
              "used": 23,
            },
            "displayName": "Org Outside J1",
            "id": "75f092de-c730-4a62-a1dd-4324cb0b83af",
            "isMultipleDataLocationsForServicesEnabled": null,
            "marketingNotificationEmails": Array [],
            "onPremisesLastSyncDateTime": null,
            "onPremisesSyncEnabled": null,
            "postalCode": null,
            "preferredLanguage": "en",
            "privacyProfile": null,
            "provisionedPlans": Array [],
            "securityComplianceNotificationMails": Array [],
            "securityComplianceNotificationPhones": Array [],
            "state": null,
            "street": null,
            "technicalNotificationMails": Array [
              "michael.knoedel_adamjupiteronehotmailcom.onmicrosoft.cBQWNE#EXT#@fakedomainfortesting.onmicrosoft.com",
            ],
            "tenantType": "AAD",
            "verifiedDomains": Array [
              Object {
                "capabilities": "Email, OfficeCommunicationsOnline",
                "isDefault": true,
                "isInitial": true,
                "name": "orgoutsidej1.onmicrosoft.com",
                "type": "Managed",
              },
            ],
          },
        },
      },
    ],
    "_type": "microsoft_365_account",
    "createdOn": undefined,
    "defaultDomain": "orgoutsidej1.onmicrosoft.com",
    "displayName": "Local Integration",
    "id": "75f092de-c730-4a62-a1dd-4324cb0b83af",
    "intuneAccountId": undefined,
    "intuneSubscriptionState": undefined,
    "mobileDeviceManagementAuthority": undefined,
    "name": "Org Outside J1",
    "organizationName": "Org Outside J1",
    "verifiedDomains": Array [
      "orgoutsidej1.onmicrosoft.com",
    ],
  },
]
`;
